pallets/merkle/src/lib.rs

```diff
@@ -150,58 +154,53 @@ decl_module! {
 		#[weight = 0]
 		pub fn verify_zk_membership_proof(
 			origin,
-			group_id: u32,
+			group_id: GroupId,
 			leaf_com: Commitment,
 			path: Vec<(Commitment, Commitment)>,
 			s_com: Commitment,
 			nullifier: Data,
 			proof_bytes: Vec<u8>
 		) -> dispatch::DispatchResult {
 			// Ensure that nullifier is not used
-			ensure!(!UsedNullifiers::get(nullifier), "Nullifier already used.");
+			ensure!(!UsedNullifiers::get((group_id, nullifier)), "Nullifier already used.");
 			let tree = <Groups<T>>::get(group_id)
 				.ok_or("Invalid group id.")
 				.unwrap();
 			ensure!(tree.edge_nodes.len() == path.len(), "Invalid path length.");
 
 			let pc_gens = PedersenGens::default();
-			let bp_gens = BulletproofGens::new(2048, 1);
+			// TODO: should be able to pass number of generators
+			let bp_gens = BulletproofGens::new(4096, 1);
```

Comment:
These generators I'm pretty sure will be generated in a trusted ceremony and should likely be saved per tree or per fixed-height family of trees.